src/server/graphics/platform_probe.cpp

```diff
@@ -203,8 +224,55 @@ auto modules_for_device(
             best_modules_so_far.erase(first_dummy_platform, best_modules_so_far.end());
         }
     }
-    return best_modules_so_far;
-}
+
+    if (best_modules_so_far.empty())
+    {
+        if (!best_nested)
+        {
+            BOOST_THROW_EXCEPTION((std::runtime_error{"Failed to find any platforms for current system"}));
+        }
+        // Our *only* option is the nested platform; return that.
+        decltype(best_modules_so_far) nested_vec;
+        nested_vec.push_back(std::move(best_nested.value()));
+        return nested_vec;
+    }
+
+    switch(nested_selection)
+    {
+#ifndef __clang__
+    /* We handle both enum variants below.
+     * GCC does not use this particular UB to provide a better diagnostic,
+     * so we need to provide a default case.
+     *
+     * As normal, leave it guarded by !Clang, so our clang builds will
+     * error out at compile-time if we add a new enum variant.
+     */
+    default:
+        mir::fatal_error("Impossible TypePreference");
+        // We can't get here, either, but sadly we can't mark mir::fatal_error as [[noreturn]]
+        // to get the compiler to enforce that 🤦‍♀️
```

Comment:
Do we want to introduce emojis (or non-ASCII, for that matter)? :)

Reply:
🤷‍♀️ 😛